mavros_extras/src/plugins/odom.cpp

```diff
@@ -0,0 +1,176 @@
+/**
+ * @brief Odometry plugin
+ * @file odom.cpp
+ * @author James Goppert <james.goppert8@gmail.com>
+ *
+ * @addtogroup plugin
+ * @{
+ */
+/*
+ * Copyright 2017 James Goppert
+ *
+ * This file is part of the mavros package and subject to the license terms
+ * in the top-level LICENSE file of the mavros repository.
+ * https://github.com/mavlink/mavros/tree/master/LICENSE.md
+ */
+
+#include <mavros/mavros_plugin.h>
+#include <eigen_conversions/eigen_msg.h>
+
+#include <nav_msgs/Odometry.h>
+
+namespace mavros {
+namespace extra_plugins {
+using mavlink::common::MAV_ESTIMATOR_TYPE;
+/**
+ * @brief Odometry plugin
+ *
+ * Send odometry info
+ * to FCU position and attitude estimators
+ */
+class OdometryPlugin : public plugin::PluginBase {
+public:
+	OdometryPlugin() : PluginBase(),
+		odom_nh("~odometry"),
+		estimator_type(3)
+	{ }
+
+	void initialize(UAS &uas_)
+	{
+		PluginBase::initialize(uas_);
+
+		// general params
+		odom_nh.param("estimator_type", estimator_type, 3);	// defaulted to VIO type
+
+		ROS_DEBUG_STREAM_NAMED("odom", "Odometry: estimator type: " << utils::to_string_enum<MAV_ESTIMATOR_TYPE>(estimator_type));
+
+		// subscribers
+		odom_sub = odom_nh.subscribe("odom", 10, &OdometryPlugin::odom_cb, this);
+	}
+
+	Subscriptions get_subscriptions()
+	{
+		return { /* Rx disabled */ };
+	}
+
+private:
+	ros::NodeHandle odom_nh;
+	ros::Subscriber odom_sub;
+
+	int estimator_type;
+
+	const Eigen::Matrix3d ZEROM3D = Eigen::Matrix3d::Zero();
+
+	/* -*- callbacks -*- */
+
+	void odom_cb(const nav_msgs::Odometry::ConstPtr &odom)
+	{
+		Eigen::Affine3d tr;
+		Eigen::Vector3d lin_vel_enu;
+		Eigen::Vector3d ang_vel_enu;
+		tf::poseMsgToEigen(odom->pose.pose, tr);
+		tf::vectorMsgToEigen(odom->twist.twist.linear, lin_vel_enu);
+		tf::vectorMsgToEigen(odom->twist.twist.angular, ang_vel_enu);
+
+		// get current ENU FCU orientation
+		Eigen::Quaterniond enu_orientation;
+		tf::quaternionMsgToEigen(m_uas->get_attitude_orientation(), enu_orientation);
+
+		// Build 9x9 covariance matrix to be transformed and sent
+		ftf::Covariance9d cov_full;
+		ftf::EigenMapCovariance9d cov_full_map(cov_full.data());
+		ftf::EigenMapConstCovariance6d cov_in(odom->pose.covariance.data());
+		cov_full_map << cov_in, Eigen::MatrixXd::Zero(6, 3),
+		Eigen::MatrixXd::Zero(3, 6), Eigen::Matrix3d::Identity();
```

Comment:
The mavlink msg has a URT of a 9d covariance matrix (position, orientation and acceleration)